scrapy/pqueues.py

```diff
@@ -0,0 +1,218 @@
+import hashlib
+import logging
+from six import text_type
+from six.moves.urllib.parse import urlparse
+
+from queuelib import PriorityQueue
+
+from scrapy.core.downloader import Downloader
+from scrapy.http import Request
+from scrapy.signals import request_reached_downloader, response_downloaded
+
+
+logger = logging.getLogger(__name__)
+
+
+SCHEDULER_SLOT_META_KEY = Downloader.DOWNLOAD_SLOT
+
+
+def _get_from_request(request, key, default=None):
+    if isinstance(request, dict):
+        return request.get(key, default)
+
+    if isinstance(request, Request):
+        return getattr(request, key, default)
+
+    raise ValueError('Bad type of request "%s"' % (request.__class__, ))
+
+
+def _scheduler_slot_read(request, default=None):
+    meta = _get_from_request(request, 'meta', dict())
+    slot = meta.get(SCHEDULER_SLOT_META_KEY, default)
+    return slot
+
+
+def _scheduler_slot_write(request, slot):
+    meta = _get_from_request(request, 'meta', None)
+    if not isinstance(meta, dict):
+        raise ValueError('No meta attribute in %s' % (request, ))
+    meta[SCHEDULER_SLOT_META_KEY] = slot
+
+
+def _scheduler_slot(request):
+
+    slot = _scheduler_slot_read(request, None)
+    if slot is None:
+        url = _get_from_request(request, 'url')
+        slot = urlparse(url).hostname or ''
+        _scheduler_slot_write(request, slot)
+
+    return slot
+
+
+def _pathable(x):
+    pathable_slot = "".join([c if c.isalnum() or c in '-._' else '_' for c in x])
+
+    """
+        as we replace some letters we can get collision for different slots
+        add we add unique part
+    """
+    unique_slot = hashlib.md5(x.encode('utf8')).hexdigest()
+
+    return '-'.join([pathable_slot, unique_slot])
+
+
+class PrioritySlot:
+    __slots__ = ('priority', 'slot')
+
+    def __init__(self, priority=0, slot=None):
+        self.priority = priority
+        self.slot = slot
+
+    def __hash__(self):
+        return hash((self.priority, self.slot))
+
+    def __eq__(self, other):
+        return (self.priority, self.slot) == (other.priority, other.slot)
+
+    def __lt__(self, other):
+        return (self.priority, self.slot) < (other.priority, other.slot)
+
+    def __str__(self):
+        return '_'.join([text_type(self.priority),
+                         _pathable(text_type(self.slot))])
+
+
+class PriorityAsTupleQueue(PriorityQueue):
+    """
+        Python structures is not directly (de)serialized (to)from json.
+        We need this modified queue to transform custom structure (from)to
+        json serializable structures
+    """
```

Comment:
We usually use a different indentation, i.e. text inside a docstring is not indented further than quotes.